File: scripts/convert_to_js.js
const fs = require("fs");
const path = require("path");

const roots = [
  path.join(__dirname, "..", "src"),
  path.join(__dirname, "..", "client", "src"),
];
const skipNames = ["vite-env.d.ts"];

const replacements = [
  { pattern: /^import type .*\n?/gm, replace: "" },
  {
    pattern: /\bimport \{\s*type\s+([^}]+)\s*\}\s*from\s*['"][^'"]+['"];?/gm,
    replace: "import { $1 } from $2;",
  },
  { pattern: /\bexport\s+type\s+\w+\s*=\s*\{[\s\S]*?\};?/gm, replace: "" },
  { pattern: /\btype\s+\w+\s*=\s*\{[\s\S]*?\};?/gm, replace: "" },
  { pattern: /\bexport\s+interface\s+\w+\s*\{[\s\S]*?\}/gm, replace: "" },
  { pattern: /\binterface\s+\w+\s*\{[\s\S]*?\}/gm, replace: "" },
  { pattern: /useState<[^>]+>/g, replace: "useState" },
  { pattern: /useParams<[^>]+>\(\)/g, replace: "useParams()" },
  { pattern: /React\.FormEvent<[^>]+>/g, replace: "React.FormEvent" },
  { pattern: /<[^>]+>\(/g, replace: "(" },
  { pattern: /\s+as\s+[^\s;\)]+/g, replace: "" },
  { pattern: /:\s*[^=;,)]+(?=[,\)\];])/g, replace: "" },
  {
    pattern:
      /import\s+\{\s*createClient\s*\}\s+from\s+['"]@supabase\/supabase-js['"];/g,
    replace: "import { createClient } from '@supabase/supabase-js';",
  },
  { pattern: /\.tsx/g, replace: ".jsx" },
  { pattern: /\.ts/g, replace: ".js" },
];

function processFile(filePath) {
  let source = fs.readFileSync(filePath, "utf8");
  replacements.forEach(({ pattern, replace }) => {
    source = source.replace(pattern, replace);
  });
  source = source.replace(/\n{3,}/g, "\n\n");
  source = source.replace(/\n\s*\n\s*\n/g, "\n\n");
  fs.writeFileSync(filePath, source, "utf8");
  const targetPath = filePath.replace(
    /\.tsx?$/,
    filePath.endsWith(".tsx") ? ".jsx" : ".js",
  );
  fs.renameSync(filePath, targetPath);
  console.log(`Converted ${filePath} -> ${targetPath}`);
}

function walk(root) {
  if (!fs.existsSync(root)) return;
  const entries = fs.readdirSync(root, { withFileTypes: true });
  for (const entry of entries) {
    const fullPath = path.join(root, entry.name);
    if (entry.isDirectory()) {
      walk(fullPath);
    } else if (/\.tsx?$/.test(entry.name) && !skipNames.includes(entry.name)) {
      processFile(fullPath);
    }
  }
}

for (const root of roots) {
  walk(root);
}

// Convert Vite config files and tsconfig files
const extraFiles = [
  path.join(__dirname, "..", "vite.config.ts"),
  path.join(__dirname, "..", "client", "vite.config.ts"),
  path.join(__dirname, "..", "tsconfig.json"),
  path.join(__dirname, "..", "client", "tsconfig.json"),
  path.join(__dirname, "..", "src", "vite-env.d.ts"),
  path.join(__dirname, "..", "client", "src", "vite-env.d.ts"),
];

for (const filePath of extraFiles) {
  if (!fs.existsSync(filePath)) continue;
  if (filePath.endsWith(".ts") && filePath.includes("vite.config")) {
    let source = fs.readFileSync(filePath, "utf8");
    source = source.replace(
      /import \{ defineConfig \} from \"vite\";/g,
      "const { defineConfig } = require('vite');",
    );
    source = source.replace(
      /import react from \"@vitejs\/plugin-react\";/g,
      "const react = require('@vitejs/plugin-react');",
    );
    source = source.replace(
      /export default defineConfig\(/g,
      "module.exports = defineConfig(",
    );
    const targetPath = filePath.replace(/\.ts$/, ".js");
    fs.writeFileSync(targetPath, source, "utf8");
    fs.unlinkSync(filePath);
    console.log(`Converted ${filePath} -> ${targetPath}`);
  } else if (
    filePath.endsWith("tsconfig.json") ||
    filePath.endsWith("vite-env.d.ts")
  ) {
    fs.unlinkSync(filePath);
    console.log(`Removed ${filePath}`);
  }
}

File: scripts/convert_to_js.py
import os
import re

root = os.path.join(os.path.dirname(__file__), "..", "src")

patterns = [
    (re.compile(r"^import type .*\n", flags=re.MULTILINE), ""),
    (re.compile(r"\bexport\s+type\s+\w+\s*=\s*\{[^}]*\};?", flags=re.DOTALL), ""),
    (re.compile(r"\btype\s+\w+\s*=\s*\{[^}]*\};?", flags=re.DOTALL), ""),
    (re.compile(r"\bexport\s+interface\s+\w+\s*\{[^}]*\}", flags=re.DOTALL), ""),
    (re.compile(r"\binterface\s+\w+\s*\{[^}]*\}", flags=re.DOTALL), ""),
    (re.compile(r"useState<[^>]+>"), "useState"),
    (re.compile(r"useParams<[^>]+>\(\)"), "useParams()"),
    (re.compile(r"React\.FormEvent<[^>]+>"), "React.FormEvent"),
    (re.compile(r"\bconst\s+(\w+)\s*:\s*[^=;]+\s*=\s*"), r"const \1 = "),
    (re.compile(r"\blet\s+(\w+)\s*:\s*[^=;]+\s*=\s*"), r"let \1 = "),
    (re.compile(r"\bvar\s+(\w+)\s*:\s*[^=;]+\s*=\s*"), r"var \1 = "),
    (re.compile(r"\bconst\s+(\w+)\s*:\s*[^;]+;"), r"const \1;"),
    (re.compile(r"\blet\s+(\w+)\s*:\s*[^;]+;"), r"let \1;"),
    (re.compile(r"\bvar\s+(\w+)\s*:\s*[^;]+;"), r"var \1;"),
    (re.compile(r"function\s+(\w+)\s*\(([^)]*)\)\s*:\s*[^\s{]+"), r"function \1(\2)"),
    (re.compile(r"\)\s*:\s*[^\s{]+\s*\{"), ") {"),
    (re.compile(r"\s+as\s+[^\s;\)]+"), ""),
    (re.compile(r"\s*:\s*[^=;\),]+(?=[\),;])"), ""),
]

for dirpath, _, filenames in os.walk(root):
    for filename in filenames:
        if filename in ("vite-env.d.ts",):
            continue
        if filename.endswith((".ts", ".tsx")):
            path = os.path.join(dirpath, filename)
            with open(path, "r", encoding="utf-8") as f:
                source = f.read()
            result = source
            for pattern, repl in patterns:
                result = pattern.sub(repl, result)
            # Remove excess blank lines from type removal
            result = re.sub(r"\n{3,}", "\n\n", result)
            target = path[:-3] + (".jsx" if filename.endswith(".tsx") else ".js")
            with open(target, "w", encoding="utf-8") as f:
                f.write(result)
            os.remove(path)
            print(f"Converted {path} -> {target}")

# Convert vite config if ts
vite_path = os.path.join(os.path.dirname(__file__), "..", "vite.config.ts")
if os.path.exists(vite_path):
    with open(vite_path, "r", encoding="utf-8") as f:
        cfg = f.read()
    cfg = cfg.replace(
        'import { defineConfig } from "vite";\nimport react from "@vitejs/plugin-react";\n',
        "const { defineConfig } = require('vite');\nconst react = require('@vitejs/plugin-react');\n",
    )
    cfg = cfg.replace("export default defineConfig(", "module.exports = defineConfig(")
    js_path = vite_path[:-3] + ".js"
    with open(js_path, "w", encoding="utf-8") as f:
        f.write(cfg)
    os.remove(vite_path)
    print(f"Converted {vite_path} -> {js_path}")

# Remove tsconfig and vite-env if present
for extra in ["tsconfig.json", os.path.join(root, "vite-env.d.ts")]:
    if os.path.exists(extra):
        os.remove(extra)
        print(f"Removed {extra}")
